libs/client-api/src/v2/README.md

```diff
@@ -0,0 +1,929 @@
+    - Server broadcasts newly received updates to all other clients
+    - All clients converge to the same state [A:2, B:2]
+
+The offline editing capability is made possible by:
+
+- Local-first edits with CRDT data structures
+- State vector tracking to identify precisely what updates are missing
+- Two-way synchronization on reconnection
+- The `WorkspaceController` client implementation that manages connection state
+
+## Resilience Mechanisms
+
+The AppFlowy collaboration system implements several mechanisms to ensure reliability and consistency across various
+challenging scenarios.
+
+### Resource ID (Rid) Management
+
+```mermaid
+flowchart TD
+    subgraph "Rid Structure"
+        R[Resource ID]
+        R --> TS[timestamp: i64]
+        R --> SN[seq_no: u16]
+    end
+
+    subgraph "Comparison Logic"
+        CL[Rid Comparison]
+        CL --> T1{timestamp equal?}
+        T1 -->|Yes| S1[Compare seq_no]
+        T1 -->|No| T2[Compare timestamp]
+    end
+
+    subgraph "Client-Side Processing"
+        CM[Receive Update]
+        CM --> CR{Is Rid > current?}
+        CR -->|Yes| UR[Update last_message_id]
+        CR -->|No| KR[Keep current last_message_id]
+    end
+
+    style R fill: #f9f, stroke: #333
+    style CL fill: #ffe6cc, stroke: #333
+    style CM fill: #d1f0ff, stroke: #333
+```
+
+Resource IDs (Rid) provide a robust ordering mechanism for updates, ensuring consistency even when messages arrive out
+of order:
+
+```rust
+pub struct Rid {
+    pub timestamp: i64,  // UNIX timestamp when update was created
+    pub seq_no: u16,     // Sequence number for same-timestamp updates
+}
+
+impl Ord for Rid {
+    fn cmp(&self, other: &Self) -> Ordering {
+        match self.timestamp.cmp(&other.timestamp) {
+            Ordering::Equal => self.seq_no.cmp(&other.seq_no),
+            ordering => ordering,
+        }
+    }
+}
+```
+
+#### Detailed Rid Implementation and Purpose
+
+The Rid mechanism is a critical component for maintaining causal consistency in the distributed CRDT system. It provides
+several key functionalities:
+
+1. **Message Identification**: Each update is assigned a unique Rid that contains:
+    - A timestamp (Unix milliseconds) when the update was created
+    - A sequence number to distinguish between updates created at the same millisecond
+
+2. **Total Ordering**: Rids establish a clear, total ordering of all updates across the system:
+   ```rust
+   // In libs/appflowy-proto/src/shared/rid.rs
+   impl PartialOrd for Rid {
+     fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+       Some(self.cmp(other))
+     }
+   }
+   
+   impl Ord for Rid {
+     fn cmp(&self, other: &Self) -> Ordering {
+       match self.timestamp.cmp(&other.timestamp) {
+         Ordering::Equal => self.seq_no.cmp(&other.seq_no),
+         ordering => ordering,
+       }
+     }
+   }
+   ```
+
+3. **Causal History Tracking**: By tracking the last received Rid (`last_message_id`), the system can:
```

Comment:
Rid is useful for 2 reasons:
1. So that we can quickly check for new updates if client has disconnected for a second.
2. Since it has a timestamp component, we can also check in the database if new collabs were created since the last sync time (last Rid received).